File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章技能升级.cpp
//技能升级 2022年第十三届省赛C/C++大学C组，Java研究生组，Python大学B组、研究生组
// lanqiaoOJ 2129

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
const int N = 1e5+4;
ll n, m;
ll a[N] = {0}, b[N] = {0};
// 最后一次加攻击能不能到x
ll check(ll x) {
    ll cnt = 0;
    for(int i=0; i<n; i++) {
        if (a[i] >= x) {
            cnt += (a[i] - x) / b[i] + 1;
            if (cnt >= m) {
                return 1;
            }
        }
    }
    return 0;
}
// 二分枚举最后一次攻击力最高能加多少
int getHighestLast() {
    int low = 0, high = 1e6;
    while(low < high) {
        int mid =  (low + high + 1) >> 1;
        if (check(mid)) { // 如果mid可以
            low = mid;
        } else {
            high = mid-1;
        }
    }
    return low;
}
int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    cin >> n >> m;
    ll totalCnt = 0;
    for(int i=0; i<n; i++) {
        cin >> a[i] >> b[i];
        if (a[i] % b[i] == 0) {
            totalCnt += a[i] / b[i];
        } else {
            totalCnt += a[i] / b[i] + 1;
        }
    }
    m = min(m, totalCnt);
    int highestLast = getHighestLast();
    ll ans = 0;
    for (int i=0 ; i<n; i++) {
        if (a[i] >= highestLast) {
            int cnt = (a[i] - highestLast) / b[i] + 1;
            int last = a[i] - (cnt-1) * b[i];
            if (last == highestLast) {
                cnt--;
                last += b[i];
            }
            ans += (a[i] + last) * cnt >> 1;
            m -= cnt;
        }
    }
    cout << ans + m * highestLast;
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章技能升级.py
#技能升级 2022年第十三届省赛C/C++大学C组，Java研究生组，Python大学B组、研究生组
#lanqiaoOJ 2129

def check(mid):                          #最后一次技能升级，最多能不能到mid
    cnt = 0 
    for i in range(n):    
        if a[i] < mid:  continue         #第i个技能的初值还不够mid，不用这个技能
        cnt += (a[i] - mid) // b[i] + 1  #第i个技能用掉的次数
        if cnt >= m:   return True       #所有技能升级总次数大于等于m次，说明mid设小了
    return False                         #所有技能升级总次数小于m次，说明mid设大了
 
n, m = map(int, input().split())
a = []                                   #存ai
b = []                                   #存bi
for i in range(n):
    a_,b_ = map(int,input().split())
    a.append(a_)
    b.append(b_)
L,R = 1,1000000                          #二分枚举最后一次攻击力最高能加多少
while(L <= R):
    mid = (L + R) // 2
    if check(mid): L = mid + 1           #增加mid
    else:          R = mid - 1           #减小mid
attack = 0
cnt = m
for i in range(n):
    if a[i] < R:  continue
    t = (a[i] - L) // b[i] + 1           #第i个技能升级次数
    if a[i] - b[i] * (t - 1) == R:  t -= 1  #这个技能每次升级刚好等于R，其他技能更好
    attack += (a[i] * 2 - (t - 1) * b[i]) * t / 2  
    cnt -= t
print(int(attack) + cnt * R)

